### ISSUE TITLE
When FQDN policy is applied, DNS A Records with more than 24 IPs time out

### ISSUE THREAD

**Issue by Starblade42** [cilium/cilium#9755]
# Bug Report
## Current Behavior
When the test record `deleteme.jhuff-k8s-04.ethos.adobe.net` has `25+` IP's in it, and I apply policy that uses FQDN DNS policy (even `rules.dns{'matchPattern':'*'}`, I get DNS lookup failure with TIMEOUT:

```
root@bash-test-8666b76bdd-nv2bp:/# dig deleteme.jhuff-k8s-04.ethos.adobe.net +short

; <<>> DiG 9.11.3-1ubuntu1.11-Ubuntu <<>> deleteme.jhuff-k8s-04.ethos.adobe.net +short
;; global options: +cmd
;; connection timed out; no servers could be reached
```

### If I delete ONE A record for the domain name (from 25 IPs to 24 IPs) , without changing the DNS policy, the name resolves just fine:
```
root@bash-test-8666b76bdd-nv2bp:/# dig deleteme.jhuff-k8s-04.ethos.adobe.net

; <<>> DiG 9.11.3-1ubuntu1.11-Ubuntu <<>> deleteme.jhuff-k8s-04.ethos.adobe.net
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 3726
;; flags: qr rd ra; QUERY: 1, ANSWER: 25, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
;; QUESTION SECTION:
;deleteme.jhuff-k8s-04.ethos.adobe.net. IN A

;; ANSWER SECTION:
deleteme.jhuff-k8s-04.ethos.adobe.net. 30 IN A	52.254.107.116
deleteme.jhuff-k8s-04.ethos.adobe.net. 30 IN A	20.190.250.115
deleteme.jhuff-k8s-04.ethos.adobe.net. 30 IN A	52.254.107.112
deleteme.jhuff-k8s-04.ethos.adobe.net. 30 IN A	52.254.107.113
deleteme.jhuff-k8s-04.ethos.adobe.net. 30 IN A	20.190.250.118
deleteme.jhuff-k8s-04.ethos.adobe.net. 30 IN A	52.254.107.121
deleteme.jhuff-k8s-04.ethos.adobe.net. 30 IN A	52.254.107.122
deleteme.jhuff-k8s-04.ethos.adobe.net. 30 IN A	52.254.107.120
deleteme.jhuff-k8s-04.ethos.adobe.net. 30 IN A	20.190.250.113
deleteme.jhuff-k8s-04.ethos.adobe.net. 30 IN A	52.254.107.115
deleteme.jhuff-k8s-04.ethos.adobe.net. 30 IN A	52.254.107.111
deleteme.jhuff-k8s-04.ethos.adobe.net. 30 IN A	52.254.107.126
deleteme.jhuff-k8s-04.ethos.adobe.net. 30 IN A	52.254.107.123
deleteme.jhuff-k8s-04.ethos.adobe.net. 30 IN A	20.190.250.116
deleteme.jhuff-k8s-04.ethos.adobe.net. 30 IN A	52.254.107.114
deleteme.jhuff-k8s-04.ethos.adobe.net. 30 IN A	52.254.107.117
deleteme.jhuff-k8s-04.ethos.adobe.net. 30 IN A	20.190.250.117
deleteme.jhuff-k8s-04.ethos.adobe.net. 30 IN A	52.254.107.125
deleteme.jhuff-k8s-04.ethos.adobe.net. 30 IN A	20.190.250.112
deleteme.jhuff-k8s-04.ethos.adobe.net. 30 IN A	52.254.107.119
deleteme.jhuff-k8s-04.ethos.adobe.net. 30 IN A	52.254.107.127
deleteme.jhuff-k8s-04.ethos.adobe.net. 30 IN A	20.190.250.114
deleteme.jhuff-k8s-04.ethos.adobe.net. 30 IN A	52.254.107.118
deleteme.jhuff-k8s-04.ethos.adobe.net. 30 IN A	52.254.107.124
deleteme.jhuff-k8s-04.ethos.adobe.net. 30 IN A	20.190.250.119

;; Query time: 71 msec
;; SERVER: 192.168.0.2#53(192.168.0.2)
;; WHEN: Thu Dec 12 22:51:48 UTC 2019
;; MSG SIZE  rcvd: 1391
```

## Expected Behavior
The DNS record is not blocked, and the upstream resolver is able to resolve other names. I'd expect this record to resolve, no matter the number of IPs.

## Bug report

**General Information**

- Cilium version (run `cilium version`)
```
Client: 1.6.4 8048d320a 2019-11-27T17:00:12+01:00 go version go1.12.13 linux/amd64
Daemon: 1.6.4 8048d320a 2019-11-27T17:00:12+01:00 go version go1.12.13 linux/amd64
```
- Kernel version (run `uname -a`)
```
Linux vmss-agent-worker-jhuffk8s04-vtzwu00000K 4.19.84-coreos #1 SMP Tue Nov 19 20:51:06 -00 2019 x86_64 x86_64 x86_64 GNU/Linux
```
- Orchestration system version in use (e.g. `kubectl version`, Mesos, ...)
```
Client Version: version.Info{Major:"1", Minor:"16", GitVersion:"v1.16.3", GitCommit:"b3cbbae08ec52a7fc73d334838e18d17e8512749", GitTreeState:"clean", BuildDate:"2019-11-14T04:24:34Z", GoVersion:"go1.12.13", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"15", GitVersion:"v1.15.6", GitCommit:"7015f71e75f670eb9e7ebd4b5749639d42e20079", GitTreeState:"clean", BuildDate:"2019-11-13T11:11:50Z", GoVersion:"go1.12.12", Compiler:"gc", Platform:"linux/amd64"}
```
## Relevant Artifacts
- Here's our network policy:
    - https://gist.github.com/Starblade42/dae5a72c060bbc283f9cae57eabb66e4
- And here's the bash deployment I'm using for testing:
    - https://gist.github.com/Starblade42/eca3ff38f0f215f0b0950487cab76852

- I have a DNS name that is **25 IP's long**, feel free to use it to reproduce the issue: 
    - `deleteme.jhuff-k8s-04.ethos.adobe.net` (there's nothing listening on those IP's, but they do belong to my test setup).
- Here's one that is **24 IP's long** and doesn't trigger the bug:
   - `deleteme1.jhuff-k8s-04.ethos.adobe.net`
- Upload a system dump (run `curl -sLO
releases.cilium.io/tools/cluster-diagnosis.zip &&
[cilium-bugtool-20191212-211820.476+0000-UTC-482671139.tar.zip](https://github.com/cilium/cilium/files/3958145/cilium-bugtool-20191212-211820.476%2B0000-UTC-482671139.tar.zip)

python cluster-diagnosis.zip sysdump` and then attach the generated zip file)

**How to reproduce the issue**

1. Apply policy that uses FQDN DNS magickery in Cilium
2. Try to look up an IP (like `deleteme.jhuff-k8s-04.ethos.adobe.net`) that has "too many" records (>24 records). *Timeout*.
3. Try to look up a record that has <= 24 records, like `deleteme1.jhuff-k8s-04.ethos.adobe.net`. *Succeed*!

**Comment by raybejjani** [comment]
@Starblade42 Thanks for the comprehensive report! I suspect this is something with setting the truncated bit as the query returns when over TCP. Oddly, the daemon logs that it is responding to the original query.
I'll look into it further. Sorry for the inconvenience!

**Comment by cgoudie** [comment]
Could it have something to do with the udp packet size? I wonder if it's being fragmented in some environments.

**Comment by raybejjani** [comment]
Finally had a chance to look at this more. @cgoudie, you are correct! It is hitting a fragmentation error between the proxy (aka the agent) and the pod. I think I know how/why this happens but I need to check more with folks more familiar with the datapath (and how it handles UDP).
```
<- endpoint 217 flow 0xb45dc241 identity 35732->0 state new ifindex 0 orig-ip 0.0.0.0: 10.16.4.150 -> 172.20.0.10 TimeExceeded(FragmentReassemblyTimeExceeded)
xx drop (Policy denied (L3)) flow 0xb45dc241 to endpoint 0, identity 35732->2: 10.16.4.150 -> 172.20.0.10 TimeExceeded(FragmentReassemblyTimeExceeded)
```

**Comment by raybejjani** [comment]
Righto, I had a chance to look through this with someone that would know, and it is a fragmentation issue. In particular, IP NAT translation doesn't happen on both fragments of the response packet because of missing port info. The second fragment is then dropped by policy, although it probably wouldn't be re-assemble if allowed (see below) because the IPs don't match. Fixing this is non-trivlal with the current datapath implementation, I'm afraid.

The most reliable solution is to force using TCP by setting the DNS TC bit. I'm guessing that this is a service-discovery scenario, though, so that may be difficult to apply (since the source of the response may not support it, or not support setting it at the correct packet size).
A more involved fix is to:
1- Use the DNS resolver IPs directly (e.g. a headless service). This avoids the partial IP translation on the response packets.
2- Allow ingress traffic from the DNS machines. This covers any case where the second fragment is mis-identified and not reassembled correctly. This can happen if connection tracking expects responses to the pod's source port but the fragment doesn't have it. In such cases, it isn't a response and would have normal policy apply, which would otherwise drop it.

**Comment by lrouquette** [comment]
> 2- Allow ingress traffic from the DNS machines. This covers any case where the second fragment is mis-identified and not reassembled correctly. This can happen if connection tracking expects responses to the pod's source port but the fragment doesn't have it. In such cases, it isn't a response and would have normal policy apply, which would otherwise drop it.

@raybejjani - I'm curious: what would that policy look like?

**Comment by raybejjani** [comment]
Sorry folks, I completely missed your question :/ An answer is below but https://github.com/cilium/cilium/pull/10366 might be the reason you were seeing issues. DNS compression was never enabled, so DNS responses would not be compressed between the proxy and pods. The compression removes a lot of duplication which, with A lookups, is significant since each IP has the domain listed next to it. My testing with the example domain we had here shows it compressing to <512B which should sail though MTU restrictions.

A policy to allow fragments would be something as simple as:
```
apiVersion: "cilium.io/v2"
kind: CiliumNetworkPolicy
metadata:
  name: "allow-dns-response-fragments"
spec:
  endpointSelector:
    matchLabels:
      name: mypod
  ingress:
  - fromEndpoints:
     - matchLabels:
         "k8s:io.kubernetes.pod.namespace": kube-system
         "k8s:k8s-app": kube-dns
```
but you can include source port information in the fromEndpoints section as well (note, there is no ingress DNS proxying so there are no L7 rules). The idea with this is to treat the fragments as allowed ingress traffic. This way, the packets eventually arrive at the pod.

**Comment by joestringer** [comment]
^ Just to clarify above, for existing releases the inclusion of ports in the policy is the tricky part that causes fragments to be dropped, because packets after the first do not include the port information. If the policy remains at L3 (eg, pod labels) then fragments will be allowed through.

And as a status update, there's also ongoing work for v1.8 to support fragments with L4 policies as well (#10076)

**Comment by tgraf** [comment]
This has been fixed with fragment support introduced by #10076.

The fragmentation support is too big to backport so I'm closing this for now. Feel free to re-open if this is not desirable for anybody looking at this issue and we can explore other options.
